src/plugins/clips-executive/clips/resource-locks.clp

```diff
@@ -34,61 +34,85 @@
 )
 
 (defrule resource-locks-request-lock
+  "Request a mutex for each required resource if none of the respective mutexes
+   is already locked or has a pending request."
   (goal (mode COMMITTED)
-        (acquired-resources $?acq)
-        (required-resources $?req))
-  ; We may have multiple goals requiring the same resource. If that's the case,
-  ; wait until the other goal releases the resources.
+        (id ?goal-id)
+        (acquired-resources)
+        (required-resources $?req&:(> (length$ ?req) 0)))
   (not (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
               (request ~NONE)))
   (not (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
               (state LOCKED)))
   =>
-  (foreach ?res (set-diff ?req ?acq)
+  (foreach ?res ?req
     (printout warn "Locking resource " ?res crlf)
     (mutex-try-lock-async (resource-to-mutex ?res))
+    (assert (resource-locks-resource-requested-for ?res ?goal-id))
   )
 )
 
-(defrule resource-locks-fast-reject-goal
-  "If a resource is locked by someone else, we have not acquired any resources,
-   and we have no pending requests, then we can directly reject the goal."
+(defrule resource-locks-fast-reject-goal-locked-by-other-agent
+  "If a resource already locked by someone else, we have not acquired any
+   resources, and we have no pending requests, then we can directly reject the
+   goal."
   (wm-fact (key cx identity) (value ?identity))
   ?g <- (goal (mode COMMITTED)
               (id ?goal-id)
               (acquired-resources)
               (required-resources $?req))
+  ; The mutex is locked and there is no unprocessed request, which means that
+  ; it was either locked by someone else or for a different goal.
   (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
-         (state LOCKED) (locked-by ?locker&~?identity))
-  (not (mutex (name ?n&:(member$ (mutex-to-resource ?n) ?req))
+         (state LOCKED) (request NONE) (locked-by ?locker&~?identity))
+  (not (mutex (name ?n1&:(member$ (mutex-to-resource ?n1) ?req))
               (request ~NONE)))
   =>
   (printout warn "Rejecting goal " ?goal-id ", " (mutex-to-resource ?n)
                  " is already locked by " ?locker crlf)
   (modify ?g (mode FINISHED) (outcome REJECTED))
 )
 
-(defrule resource-locks-lock-acquired
-  ?m <- (mutex (name ?res) (request LOCK) (response ACQUIRED))
+(defrule resource-locks-fast-reject-goal-locked-for-another-goal
+  "If a resource already locked for another goal, we have not acquired any
+   resources, and we have no pending requests, then we can directly reject the
+   goal."
   ?g <- (goal (mode COMMITTED)
+              (id ?goal)
+              (acquired-resources)
```

Comment:
What happens if the goal already got a resource and the other goal has a resource that the goal is waiting on? Then, according to this rule, the goal would not be fast rejected, but the mutex would not be rejected (I suppose), thus the goal would not be rejected as well. As far as I understand, in this case the goal would wait for the other goal to finish, which is exactly what we dont want to happen.

Reply:
Or is this handled by the guard of resource-locks-request-lock, that there is no request pending?